File: python/Data Structures and Algorithms in Python Book/oop/geometricprogression.py
from progression import Progression

class GeometricProgression(Progression):
    
    def __init__(self, start=1, base = 2):
        super().__init__(start=start)
        self._base = base
    
    def _advance(self):
       self._current *= self._base


if __name__ == "__main__":
    geometric_progression = GeometricProgression(start = 3, base = 2)
    geometric_progression.print_progression(20)

File: java/ood/GeometricProgression.java
package ood;

public class GeometricProgression extends Progression {
    // store the base / factor
    protected long base;

    /** Connstruct progression 1, 2, 4, 8, ... */
    public GeometricProgression() {
        this(1, 2);                                         // start at 1 with base of 2
    }

    /** Construct progression 1, b, b^2, b^3, b^4, ... for base of b*/
    public GeometricProgression(long base) {
        this(1, base);
    }

    /** Construct geometric progression with arbitrary start and base */
    public GeometricProgression(long start, long base) {
        super(start);
        this.base = base;
    }

    @Override
    protected void advance() {
        current *= base;                                // multiply current by the geometric base
    }

}
